[1LP][RFR] Automate test: Create schedule for base report

valaparthvi opened the pull request:
This PR adds a test case, which creates a schedule for a base report and asserts the schedule was created without any error.

{{pytest: cfme/tests/intelligence/reports/test_crud.py::test_reports_crud_schedule_for_base_report_once --use-template-cache -sqvvv}}

valaparthvi commented on cfme/tests/intelligence/reports/test_crud.py:
```diff
@@ -264,3 +264,23 @@ def test_reports_delete_saved_report(appliance, request):
+    view.configuration.item_select("Add a new Schedule")
+    add_view = report.create_view(NewScheduleView)
+    data = {"hour": "12", "minute": "10"}
+    add_view.fill(data)
```
When I go to the Details page of a report and click on 'Add new Schedule', it opens up the form with some details pre-filled, I want to preserve those details and only make a few changes, before adding the schedule. Using the `create` method would direct to 'Add new Schedule' page and load an empty form.

valaparthvi replied:
And I am only creating a schedule here, not deleting it, so I don't think it's CRUD operation. Although, I could change the test name to something shorter.

digitronik replied:
better you can import this `NewScheduleView` in `report.py`. create a method in `Report` class.

valaparthvi commented:
@digitronik 
I have made the changes you requested.
1) Add a `ReportSchedule` view, which can be used to add a schedule from report's `Details` page.
2) Add a `create_schedule` method, which creates a schedule for the base report.
3) Modify test in accordance with the new changes.

digitronik commented on cfme/intelligence/reports/reports.py:
```diff
@@ -312,6 +313,42 @@ def delete(self, cancel=False):
     def saved_reports(self):
         return self.collections.saved_reports
 
+    def create_schedule(self, name=None, description=None, active=None,
```
Required:
Why all argument default value `None`?

valaparthvi replied:
That is in the case when no argument value is passed in the data.

digitronik commented on cfme/intelligence/reports/reports.py:
```diff
@@ -312,6 +313,42 @@ def delete(self, cancel=False):
+        view.flash.assert_no_error()
+
+        schedule = self.appliance.collections.schedules.instantiate(
+            name=name or self.menu_name,
```
Question:
can you please explain this?

valaparthvi replied:
This instantiates the newly created schedule and returns it.

digitronik commented on cfme/tests/intelligence/reports/test_crud.py:
```diff
@@ -264,3 +263,26 @@ def test_reports_delete_saved_report(appliance, request):
     assert not report.exists
+
+
+def test_reports_create_schedule_for_base_report_once(appliance, request):
```
Required:
This should be a crud operation.
Please create and delete this scheduled report properly and assert for both
remove finalizer

digitronik commented on cfme/intelligence/reports/reports.py:
```diff
@@ -556,6 +593,20 @@ def step(self):
+    def step(self):
+        self.prerequisite_view.reports.tree.click_path(
+            "All Reports",
+            self.obj.type or self.obj.company_name,
```
I am suspect about these or. are those required?

valaparthvi replied:
It is required to navigate to the details page of the report and then select `Add a new Schedule` from the configuration menu.

valaparthvi commented:
@digitronik I have added a new `ReportScheduleView`, to fix the failing `is_displayed` method. I had done a few things incorrectly, but have now fixed it.

mshriver commented on cfme/intelligence/reports/reports.py:
```diff
@@ -225,6 +226,24 @@ def is_displayed(self):
         )
 
 
+class ReportScheduleView(SchedulesFormCommon):
```
Hang on I'm pretty sure all of this already exists in the schedules.py module:
https://github.com/ManageIQ/integration_tests/blob/master/cfme/intelligence/reports/schedules.py

mshriver commented on cfme/tests/intelligence/reports/test_crud.py:
```diff
@@ -264,3 +263,26 @@ def test_reports_delete_saved_report(appliance, request):
     assert not report.exists
+
+
+def test_reports_crud_schedule_for_base_report_once(appliance, request):
```
I think this case is covered by test_schedule_queue?  

https://github.com/ManageIQ/integration_tests/blob/master/cfme/tests/intelligence/reports/test_reports_schedules.py

This uses several schedules from the cfme/data/ directory.

valaparthvi commented:
Hey @mshriver, we have views and objects to create a completely new schedule, but as far as I know, there is no way to create a schedule for an existing report. Take this for example:
1) Navigate to a report's `Details` page.
2) Click on `Configuration` and select `Add a new schedule`.
3) You'll see a form with some pre-filled data such as `Filters`, `Name` and `Description`.

Take a look at [this](https://up1-vrutkovs.int.open.paas.redhat.com/#oyZ7jTaXqjU8xs6ig-Uv_g).
I don't think we have any view or object that can directly create a schedule for a report like this.

And the test case is similar to `test_schedule_queue`, but `test_reports_crud_schedule_for_base_report_once` uses the new view and method.

But, there is one drawback I hadn't noticed before. The method `create_schedule` will fail if no data is passed since the `Add` button on New Schedule page does not activate until and unless some change is made in the form.

Let me know what you think.
